internal/resolver/dns/dns_resolver.go

```diff
@@ -183,6 +191,60 @@ type dnsResolver struct {
 	// has data race with replaceNetFunc (WRITE the lookup function pointers).
 	wg                   sync.WaitGroup
 	disableServiceConfig bool
+
+	polling chan struct{}
+}
+
+// poll begins or ends asynchronous polling of the resolver based on whether
+// err is ErrBadResolverState.
+func (d *dnsResolver) poll(errorFromClientConn error) {
+	// The only time to stop polling is when we don't receive an ErrBadResolverState from client.
+	if errorFromClientConn != balancer.ErrBadResolverState {
+		// Stop polling
+		if d.polling != nil {
+			close(d.polling)
+			d.polling = nil
+		}
+		return
+	}
+	// The goroutine that is running poll() is already running.
+	if d.polling != nil {
+		return
+	}
+	p := make(chan struct{})
+	d.polling = p
+	// This exponential backoff go routine will be running at most once.
+	d.wg.Add(1)
+	go func() {
+		defer d.wg.Done()
+		// Exponentially backoff until it hopefully succeeds.
+		for i := 0; ; i++ {
+			t := newTimer(backoff.DefaultExponential.Backoff(i))
+			select {
+			case <-p:
+				// Polling was successful.
+				t.Stop()
+				return
+			case <-d.ctx.Done():
+				return
+			case <-t.C:
+				select {
+				case <-p:
+					// Polling was successful.
+					return
+				default:
+				}
+				// Timer expired; re-resolve.
+			}
+			// After select for testing purposes. Forcing a re-resolution would cause many tests to be flaky.
+			// This allows a re-resolution to be guarded by the resolveNowBackoff() function, which can be toggled
+			// in a test.
+			select {
+			case d.pn <- struct{}{}:
+			default:
+			}
```

Comment:
Your comment: "Instead of calling yourself with ResolveNow, can the polling be integrated into watcher instead? E.g. another case in the existing select?". I thought this implied creating another channel. I can switch this to calling d.rn instead, but I'll wait for your thoughts.

Reply:
I thought about this for a bit. I think your first pass you had a problem with the call ResolveNow which would then signal rn channel. Your second pass you seem to have a problem with adding an extra pn channel when there's already a rn channel. I'll try and make the best of both worlds and call rn directly.

Reply:
Oh, right: in the first round of comments, I was suggesting integrating this whole polling mechanism into the watcher loop instead.  That would mean not having this function at all, except as a way to tell the watcher to begin/end polling.  See if you can do that.  If that doesn't work well, then when it's time to do a resolution, calling `ResolveNow` makes the most sense.

Reply:
I've been thinking as criticially as my l3 mind can think about this
So it's kind of weird to me as well to spawn off from watcher() something that will call itself and have a seperate goroutine

Reply:
However, logically, I personally think that to keep it the same as it was before (key was intertwining polling re-resolutions with any ResolveNow{} calls from around the codebase) with regards to this comment: https://github.com/grpc/grpc-go/blob/04a4145cb222e042cd664db86b8c8b394e095735/internal/resolver/dns/dns_resolver.go#L277

Reply:
> So it's kind of weird to me as well to spawn off from watcher() something that will call itself and have a seperate goroutine

This isn't what I meant.

`watcher` is a goroutine whose job it is to wake up when something happens and resolve addresses.  Before your change, it wakes up only when `ResolveNow` is called.  But there's no reason we can't add a polling timer to the `select` it already has.  My suggestions is to change it to:

```go
func Build() {
	d := &dnsResolver{
		pollTimer: time.NewTimer(0)
	}
	// No longer need ResolveNow() call in Build, as pollTimer fires immediately
}

func (d *dnsResolver) watcher() {
	defer d.wg.Done()
	nextRate := <minimum backoff interval>
	for {
		select {
		case <-d.ctx.Done():
			return
		case <-d.rn:
		case <-d.pollTimer.C:
		}

		state, err := d.lookup()
		if err != nil {
			d.cc.ReportError(err)
		} else {
			err = d.cc.UpdateState(*state)
		}
		if err == nil {
			// Success; stop polling.
			if !d.pollTimer.Stop() {
				<-d.pollTimer.C
			}
			nextRate := <minimum backoff interval>
		} else {
			d.pollTimer = time.NewTimer(nextRate)
			// adjust nextRate
		}
	}
}
```

Reply:
Moved polling from separate goroutine into watcher.